File: NavCalc/NavCalc.Core/Calculations/CurrentVector.cs
using System;
using static NavCalc.Core.Calculations.NavigationMath;

namespace NavCalc.Core.Calculations
{
    public static class CurrentVector
    {
        /// <summary>
        /// 実航針路・実航速力 (CMG/SMG)
        /// </summary>
        public static (double Cmg, double Smg) CmgSmg(
            double shipCourse, double shipSpeed, double setDir, double driftSpeed)
        {
            double sc = Rad(shipCourse), sd = Rad(setDir);
            double vx = shipSpeed * Math.Sin(sc) + driftSpeed * Math.Sin(sd);
            double vy = shipSpeed * Math.Cos(sc) + driftSpeed * Math.Cos(sd);

            double smg = Math.Sqrt(vx * vx + vy * vy);
            double cmg = Deg(Math.Atan2(vx, vy));
            if (cmg < 0) cmg += 360.0;

            return (cmg, smg);
        }

        /// <summary>
        /// 視針路・対水速力 (Course to Steer / Speed through Water)
        /// </summary>
        public static (double ShipCourse, double ShipSpeed) CourseToSteer(
            double requiredCMG, double requiredSMG, double setDir, double driftSpeed)
        {
            double cmgR = Rad(requiredCMG), sdR = Rad(setDir);
            double gx = requiredSMG * Math.Sin(cmgR);
            double gy = requiredSMG * Math.Cos(cmgR);
            double cx = driftSpeed * Math.Sin(sdR);
            double cy = driftSpeed * Math.Cos(sdR);

            double sx = gx - cx;
            double sy = gy - cy;

            double shipSpeed = Math.Sqrt(sx * sx + sy * sy);
            double shipCourse = Deg(Math.Atan2(sx, sy));
            if (shipCourse < 0) shipCourse += 360.0;

            return (shipCourse, shipSpeed);
        }

        /// <summary>
        /// 視針路・実航速力 (Course to Steer & SMG)
        /// </summary>
        public static (double ShipCourse, double Smg, bool Possible) CourseToSteerSMG(
            double requiredCMG, double shipSpeed, double setDir, double driftSpeed)
        {
            double cmgR = Rad(requiredCMG), sdR = Rad(setDir);
            double alpha = sdR - cmgR;
            double sinBeta = (driftSpeed / shipSpeed) * Math.Sin(alpha);

            if (Math.Abs(sinBeta) > 1)
                return (double.NaN, double.NaN, false);

            double beta = Math.Asin(sinBeta);
            double shipCourse = requiredCMG + Deg(beta);
            if (shipCourse < 0) shipCourse += 360.0;
            if (shipCourse >= 360) shipCourse -= 360.0;

            double gamma = Math.PI - alpha - beta;
            double sinAlpha = Math.Sin(alpha);
            double smg = shipSpeed * Math.Sin(gamma) / (Math.Abs(sinAlpha) < 1e-10 ? 1e-10 : sinAlpha);

            return (shipCourse, Math.Abs(smg), true);
        }

        /// <summary>
        /// 流向・流速 (Set & Drift)
        /// </summary>
        public static (double SetDir, double Drift) SetAndDrift(
            double shipCourse, double shipSpeed, double cmg, double smg)
        {
            double scR = Rad(shipCourse), cmgR = Rad(cmg);
            double cx = smg * Math.Sin(cmgR) - shipSpeed * Math.Sin(scR);
            double cy = smg * Math.Cos(cmgR) - shipSpeed * Math.Cos(scR);

            double drift = Math.Sqrt(cx * cx + cy * cy);
            double setDir = Deg(Math.Atan2(cx, cy));
            if (setDir < 0) setDir += 360.0;

            return (setDir, drift);
        }
    }
}

File: src/lib/utils/currentVector.ts
import { rad, deg } from "./navigationMath";

/**
 * 実航針路・実航速力 (CMG/SMG)
 * 船速+針路 と 潮流方向+速度 からベクトル加算
 */
export function cmgSmg(
  shipCourse: number,
  shipSpeed: number,
  setDir: number,
  driftSpeed: number,
) {
  const sc = rad(shipCourse),
    sd = rad(setDir);
  const vx = shipSpeed * Math.sin(sc) + driftSpeed * Math.sin(sd);
  const vy = shipSpeed * Math.cos(sc) + driftSpeed * Math.cos(sd);

  const smg = Math.sqrt(vx * vx + vy * vy);
  let cmg = deg(Math.atan2(vx, vy));
  if (cmg < 0) cmg += 360;

  return { cmg, smg };
}

/**
 * 視針路・対水速力 (Course to Steer / Speed through Water)
 * 目標CMG と 潮流 → 必要な船の針路と速力
 */
export function courseToSteer(
  requiredCMG: number,
  requiredSMG: number,
  setDir: number,
  driftSpeed: number,
) {
  // 目標 ground vector - current vector = ship vector
  const cmgR = rad(requiredCMG),
    sdR = rad(setDir);
  const gx = requiredSMG * Math.sin(cmgR);
  const gy = requiredSMG * Math.cos(cmgR);
  const cx = driftSpeed * Math.sin(sdR);
  const cy = driftSpeed * Math.cos(sdR);

  const sx = gx - cx;
  const sy = gy - cy;

  const shipSpeed = Math.sqrt(sx * sx + sy * sy);
  let shipCourse = deg(Math.atan2(sx, sy));
  if (shipCourse < 0) shipCourse += 360;

  return { shipCourse, shipSpeed };
}

/**
 * 視針路・実航速力 (Course to Steer & SMG)
 * 船速固定、目標CMG方向へ進むための視針路とSMG
 */
export function courseToSteerSMG(
  requiredCMG: number,
  shipSpeed: number,
  setDir: number,
  driftSpeed: number,
) {
  const cmgR = rad(requiredCMG),
    sdR = rad(setDir);
  // 潮流方向とCMGの角度差
  const alpha = sdR - cmgR;
  // 正弦定理で船の針路偏角を求める
  const sinBeta = (driftSpeed / shipSpeed) * Math.sin(alpha);
  if (Math.abs(sinBeta) > 1) {
    return { shipCourse: NaN, smg: NaN, possible: false };
  }
  const beta = Math.asin(sinBeta);
  let shipCourse = requiredCMG + deg(beta);
  if (shipCourse < 0) shipCourse += 360;
  if (shipCourse >= 360) shipCourse -= 360;

  // SMG = 正弦定理から
  const gamma = Math.PI - alpha - beta;
  const smg = (shipSpeed * Math.sin(gamma)) / Math.sin(alpha || 1e-10);

  return { shipCourse, smg: Math.abs(smg), possible: true };
}

/**
 * 流向・流速 (Set & Drift)
 * 船の針路・速力 と CMG/SMG → 潮流ベクトル
 */
export function setAndDrift(
  shipCourse: number,
  shipSpeed: number,
  cmg: number,
  smg: number,
) {
  const scR = rad(shipCourse),
    cmgR = rad(cmg);
  // current = ground - ship
  const cx = smg * Math.sin(cmgR) - shipSpeed * Math.sin(scR);
  const cy = smg * Math.cos(cmgR) - shipSpeed * Math.cos(scR);

  const drift = Math.sqrt(cx * cx + cy * cy);
  let setDir = deg(Math.atan2(cx, cy));
  if (setDir < 0) setDir += 360;

  return { setDir, drift };
}
